server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java

```diff
@@ -103,22 +103,63 @@ public static EsThreadPoolExecutor newScaling(
         TimeUnit unit,
         boolean rejectAfterShutdown,
         ThreadFactory threadFactory,
-        ThreadContext contextHolder
+        ThreadContext contextHolder,
+        TaskTrackingConfig config
     ) {
         ExecutorScalingQueue<Runnable> queue = new ExecutorScalingQueue<>();
-        EsThreadPoolExecutor executor = new EsThreadPoolExecutor(
+        EsThreadPoolExecutor executor;
+        if (config.trackOngoingTasks()) {
```

Comment:
This condition is checking if ongoing tasks should be [tracked in addition to the regular tracking](https://github.com/elastic/elasticsearch/blob/3cdc7086d509d3cb4e24ebcdce3082d5e8d2d796/server/src/main/java/org/elasticsearch/common/util/concurrent/TaskExecutionTimeTrackingEsThreadPoolExecutor.java#L93).

I think we want to check if the `config != DO_NOT_TRACK && config != null`?

Reply:
Hmm, sorry I'm not sure I understand. The config is non-nullable so the null check is not needed.

Why would we check the `DO_NOT_TRACK` in particular? The code you [linked to](https://github.com/elastic/elasticsearch/blob/3cdc7086d509d3cb4e24ebcdce3082d5e8d2d796/server/src/main/java/org/elasticsearch/common/util/concurrent/TaskExecutionTimeTrackingEsThreadPoolExecutor.java#L93) checks `trackOngoingTasks` (which we also do here)

Reply:
Ah, I think you're referring to the `trackExecutionTime` flag (which is only ever NOT `true` for the `DEFAULT` and `DO_NOT_TRACK` objects) and only used in `newFixed`. I'll switch the flag to check it here too instead of the `trackOngoingTasks`

Reply:
Thanks for updating this :)

>Why would we check the DO_NOT_TRACK in particular? The code you [linked to](https://github.com/elastic/elasticsearch/blob/3cdc7086d509d3cb4e24ebcdce3082d5e8d2d796/server/src/main/java/org/elasticsearch/common/util/concurrent/TaskExecutionTimeTrackingEsThreadPoolExecutor.java#L93) checks trackOngoingTasks (which we also do here)

Ah, to clarify, the code I linked showed how `trackOngoingTasks` is used: in addition to regular tracking of execution times, the `TaskExecutionTimeTrackingEsThreadPoolExecutor` can also store the state of ongoing tasks.  `trackOngoingTasks` doesn't indicate if tracking should be globally enabled or not, it determines if this extra tracking should be enabled.  Ongoing tasks haven't yet contributed to the EWMA of task execution time, as that's only recorded on task completion.  By  enabling tracking for ongoing tasks, we can get them via [getOngoingTasks()](https://github.com/salvatore-campagna/elasticsearch/blob/3cdc7086d509d3cb4e24ebcdce3082d5e8d2d796/server/src/main/java/org/elasticsearch/common/util/concurrent/TaskExecutionTimeTrackingEsThreadPoolExecutor.java#L146) in order to take their time into account when determining how busy a threadpool was over a given interval.  

---

The reason I originally suggested checking for the `DO_NOT_TRACK` instance is that it seems (at least to me) to be intended as the marker to indicate if a `TaskExecutionTimeTrackingEsThreadPoolExecutor` should not be used, whereas `trackExecutionTime()` seems like reliance on the implementation, e.g. changes to the `TaskTrackingConfig` class might make this check invalid - we might want to track ongoing tasks without checking execution time (though that is a tad contrived, I'll admit!).   However, given that `trackExecutionTime()`  seems to be [used elsewhere](https://github.com/andreidan/elasticsearch/blob/0b7114f61cc1d93652de5f11c32062f7efba8d47/server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java#L193) for this purpose,  I think what you've got now is fine :)